addons/mail/static/src/models/thread_cache/thread_cache.js

```diff
@@ -178,13 +180,34 @@ function factory(dependencies) {
          * @returns {boolean}
          */
         _computeHasToLoadMessages() {
-            return (
-                this.thread &&
-                !this.thread.isTemporary &&
-                this.threadViews.length > 0 &&
-                !this.isLoaded &&
-                !this.isLoading
-            );
+            if (!this.thread) {
+                // happens during destroy or compute executed in wrong order
+                return false;
+            }
+            const wasCacheRefreshRequested = this.isCacheRefreshRequested;
+            // mark hint as processed
+            this.update({ isCacheRefreshRequested: false });
+            if (this.thread.isTemporary) {
+                // temporary threads don't exist on the server
+                return false;
+            }
+            if (!wasCacheRefreshRequested && this.threadViews.length === 0) {
+                // don't load message that won't be used
+                return false;
+                }
+            if (!wasCacheRefreshRequested && (this.isLoaded || this.isLoading)) {
+                // avoid duplicate RPC
+                return false;
+            }
+            const isMainCache = this.thread.mainCache === this;
+            if (isMainCache && (this.isLoaded || this.isLoading)) {
+                // Ignore request on the main cache if it is already loaded or
+                // loading. Indeed the main cache is automatically sync with
+                // server updates already, so there is never a need to refresh
+                // it past the first time.
+                return false;
+            }
+            return true;
```

Comment:
The management of loaded state of cache is a mess. Refresh requested that are ignored when either loaded or loading (thankfully distinguished with `isLoadingMore`, but the design becomes extremely weak), load messages that happen both in 1st load and extra load, having particularities of main cache that are handled too deeply in code (code reads like _"do this"_ but 2 steps later it's _"nah, just ignore in that case"_, it's very confusing).

I really hope this PR is good enough for fix in 13.4, so we don't have to fix this feature again in this version.
If that's the case, then current code is somewhat bearable, but I really want to clean this mess ASAP in master.